File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounterNode(Node):
    def __init__(self):
        super().__init__("number_counter")

        self.counter_ = 0
        self.subscriber_ = self.create_subscription(
            Int64, "number", self.callback_number_received, 10)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.server_ = self.create_service(
            SetBool, "reset_counter", self.callback_reset_counter)
        self.get_logger().info("Number Counter has been started")

    def callback_number_received(self, msg):
        number_received = msg.data
        self.get_logger().info("Received: " + str(number_received))
        self.counter_ += number_received
        self.publish_count(self.counter_)

    def callback_reset_counter(self, request, response):
        if request.data:
            self.counter_ = 0
            response.success = True
            response.message = "The counter has been reset."
            self.get_logger().info("The counter has been reset.")
        else:
            response.success = False
            response.message = "The counter has not been reset."
            self.get_logger().info("The counter has not been reset.")

        return response

    def publish_count(self, num):
        msg = Int64()
        msg.data = num
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounterNode : public rclcpp::Node
{
public:
    NumberCounterNode() : Node("number_counter")
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
            "number", 10,
            std::bind(&NumberCounterNode::callbackNumberCounter, this, std::placeholders::_1));

        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);

        server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter",
                                                                         std::bind(&NumberCounterNode::callbackResetCounter, this, _1, _2));

        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");
    }

private:
    void callbackNumberCounter(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Received: %d", msg->data);
        counter_ += msg->data;
        RCLCPP_INFO(this->get_logger(), "Total: %d", counter_);
        publishNumber(counter_);
    }

    void callbackResetCounter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                              example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data)
        {
            counter_ = 0;
            response->success = true;
            response->message = "The counter has been reset.";
            RCLCPP_INFO(this->get_logger(), "The counter has been reset. Counter: %d", counter_);
        }
        else
        {
            response->success = false;
            response->message = "The counter has not been reset.";
            RCLCPP_INFO(this->get_logger(), "The counter has not been reset. Counter: %d", counter_);
        }
    }

    void publishNumber(int64_t counter)
    {
        auto msg = example_interfaces::msg::Int64();
        msg.data = std::int64_t(counter);
        publisher_->publish(msg);
    }

    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    int64_t counter_;

    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;

    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}